src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Http/HttpParser.cs

```diff
@@ -340,69 +333,81 @@ private unsafe int FindEndOfName(byte* headerLine, int length)
 
-            if (headerLine[valueEnd + 2] != ByteLF)
+            if (Unsafe.Add(ref headerLine, valueEnd + 2) != ByteLF)
             {
-                RejectRequestHeader(headerLine, length);
+                RejectRequestHeader(ref headerLine, length);
             }
-            if (headerLine[valueEnd + 1] != ByteCR)
+            if (Unsafe.Add(ref headerLine, valueEnd + 1) != ByteCR)
             {
-                RejectRequestHeader(headerLine, length);
+                RejectRequestHeader(ref headerLine, length);
             }
 
             // Skip colon from value start
             var valueStart = nameEnd + 1;
             // Ignore start whitespace
             for (; valueStart < valueEnd; valueStart++)
             {
-                var ch = headerLine[valueStart];
+                var ch = Unsafe.Add(ref headerLine, valueStart);
                 if (ch != ByteTab && ch != ByteSpace && ch != ByteCR)
                 {
                     break;
                 }
                 else if (ch == ByteCR)
                 {
-                    RejectRequestHeader(headerLine, length);
+                    RejectRequestHeader(ref headerLine, length);
                 }
             }
 
             // Check for CR in value
             var i = valueStart + 1;
-            if (Contains(headerLine + i, valueEnd - i, ByteCR))
+            if (Contains(ref Unsafe.Add(ref headerLine, i), valueEnd - i, ByteCR))
             {
-                RejectRequestHeader(headerLine, length);
+                RejectRequestHeader(ref headerLine, length);
             }
 
             // Ignore end whitespace
             for (; valueEnd >= valueStart; valueEnd--)
             {
-                var ch = headerLine[valueEnd];
+                var ch = Unsafe.Add(ref headerLine, valueEnd);
                 if (ch != ByteTab && ch != ByteSpace)
                 {
                     break;
                 }
             }
-
-            var nameBuffer = new Span<byte>(headerLine, nameEnd);
-            var valueBuffer = new Span<byte>(headerLine + valueStart, valueEnd - valueStart + 1);
+            Span<byte> nameBuffer = MakeSpan(ref headerLine, nameEnd);
+            var valueBuffer = MakeSpan(ref Unsafe.Add(ref headerLine, valueStart), valueEnd - valueStart + 1);
 
             handler.OnHeader(nameBuffer, valueBuffer);
         }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        private static unsafe bool Contains(byte* searchSpace, int length, byte value)
+        private static unsafe Span<byte> MakeSpan(ref byte data, int length)
+        {
+            // We can use this fast path on .NET Core 2.0 to make a span from ref byte
+            // return Span<byte>.DangerousCreate(null, ref data, length);
+            // For slow span, there's no way to do this because we don't have the backing object so we
+            // need to fix the data
+            fixed (byte* pointer = &data)
```

Comment:
I believe this causes a GC hole in the general case. Kestrel is fine because we use pinned memory but in general,  the GC can only track interior pointers for fast span, slow span requires the backing object otherwise it might get collected and this span would be pointing to no where...